cclib/method/ddec.py

```diff
@@ -300,6 +327,324 @@ def calculate_refcharges(self):
 
             # In DDEC6, weights of 1/3 and 2/3 are assigned for stockholder and localized charges.
             # (Equation 50 and 58 in doi: 10.1039/c6ra04656h)
-            refcharges[atomi] = (stockholdercharges[atomi] / 3.0) + (localizedcharges[atomi] * 2.0 / 3.0)
-        
+            refcharges[atomi] = (stockholdercharges[atomi] / 3.0) + (
+                localizedcharges[atomi] * 2.0 / 3.0
+            )
+
         return refcharges, localizedcharges, stockholdercharges
+
+    def condition_densities(self):
+        # Generator object to iterate over the grid
+        xshape, yshape, zshape = self.chgdensity.data.shape
+        atoms = len(self.data.atomnos)
+        indices = (
+            (i, x, y, z)
+            for i in range(atoms)
+            for x in range(xshape)
+            for y in range(yshape)
+            for z in range(zshape)
+        )
+
+        self._rho_ref = numpy.zeros((xshape, yshape, zshape))
+
+        for atomi, xindex, yindex, zindex in indices:
+            # rho_ref -- Equation 41 in doi: 10.1039/c6ra04656h
+            self._rho_ref[xindex][yindex][zindex] = (
+                self._rho_ref[xindex][yindex][zindex]
+                + self.proatom_density[atomi][self.closest_r_index[atomi][xindex][yindex][zindex]]
+            )
+
+        self._candidates_bigPsi = []
+        self._candidates_psi = []
+
+        # Initial conditions are detailed in Figure S1 in doi: 10.1039/c6ra04656h
+        psiAI = numpy.zeros_like(self.data.atomnos, dtype=float)
+        bigpsiAI = numpy.zeros_like(self.data.atomnos, dtype=float)
+        self._y_a = []
+        self._cond_density = []
+
+        for atomi in range(len(self.data.atomnos)):
+            # y_a -- equation 40 in doi: 10.1039/c6ra04656h
+            self._y_a.append(self._ya(self.proatom_density, atomi))
+            # rho_A^cond -- equation S97 in doi: 10.1039/c6ra04656h
+            self._cond_density.append(
+                self._y_a[atomi] + bigpsiAI[atomi] * numpy.sqrt(self._y_a[atomi])
+            )
+            # Monotonic Decrease Condition (as expressed in equation S99)
+            self._cond_density[atomi] = numpy.minimum.accumulate(self._cond_density[atomi])
+            # psi_A^I -- Equation S100 in doi: 10.1039/c6ra04656h
+            psiAI[atomi] = (
+                self._integrate_from_radial([self._cond_density[atomi]], [atomi])
+                - self.data.atomnos[atomi]
+                + self.refcharges[-1][atomi]
+            )
+
+            self._candidates_bigPsi.append([bigpsiAI[atomi]])
+            self._candidates_psi.append([psiAI[atomi]])
+
+            fitpsi = True  # when convergence is reached, this is modified to False.
```

Comment:
Sorry to nitpick but as I was going through and checking the equations I realized these are phi not psi. Can we change that to stay consistent with the paper?

Reply:
(This applies throughout this function)

Reply:
I've changed the variables into phi. Thanks for noting this!